core/dbt/contracts/graph/nodes.py

```diff
@@ -584,10 +620,14 @@ def build_contract_checksum(self):
             for column in sorted_columns:
                 contract_state += f"|{column.name}"
                 contract_state += str(column.data_type)
+                contract_state += str(column.constraints)
+            if self.config.materialized in ["table", "incremental"]:
```

Comment:
Can be in a follow-up issue (perhaps when we tackle https://github.com/dbt-labs/dbt-core/issues/7335), but finding a way to avoid [duplicating](https://github.com/dbt-labs/dbt-core/blob/87e25e8692197c093e9be2c0df5fe38dea48eec5/core/dbt/parser/schemas.py#L976) this hard-coded list of supported materializations would be neat!

Reply:
Hmm, looks like we're referencing `["table", "incremental"]` explicitly in this PR as well. Could we add a property/helper on the compiled node to indicate whether a materialization can enforce constraints that does this check under the hood? Something like `self.materialization_enforces_constraints`